tools/syz-cover/syz-cover.go

```diff
@@ -156,65 +156,55 @@ func main() {
 	}
 	pcs := initPCs(rg)
 	progs := []cover.Prog{{PCs: pcs}}
-	buf := new(bytes.Buffer)
 	params := cover.HandlerParams{
 		Progs: progs,
 	}
-	if *flagExportCSV != "" {
-		if err := rg.DoCSV(buf, params); err != nil {
-			tool.Fail(err)
-		}
-		if err := osutil.WriteFile(*flagExportCSV, buf.Bytes()); err != nil {
-			tool.Fail(err)
-		}
-		return
-	}
+
 	if *flagExportLineJSON != "" {
-		if err := rg.DoLineJSON(buf, params); err != nil {
-			tool.Fail(err)
-		}
-		if err := osutil.WriteFile(*flagExportLineJSON, buf.Bytes()); err != nil {
-			tool.Fail(err)
-		}
+		doReport(params, *flagExportLineJSON, rg.DoLineJSON)
 		return
 	}
 	if *flagExportJSONL != "" {
-		if err := rg.DoCoverJSONL(buf, params); err != nil {
-			tool.Fail(err)
-		}
-		if err := osutil.WriteFile(*flagExportJSONL, buf.Bytes()); err != nil {
-			tool.Fail(err)
-		}
-		return
-	}
-	if err := rg.DoHTML(buf, params); err != nil {
-		tool.Fail(err)
-	}
-	if *flagExportHTML != "" {
-		if err := osutil.WriteFile(*flagExportHTML, buf.Bytes()); err != nil {
-			tool.Fail(err)
-		}
+		doReport(params, *flagExportJSONL, rg.DoCoverJSONL)
 		return
 	}
-	fn, err := osutil.TempFile("syz-cover")
-	if err != nil {
+
+	log.Logf(1, "start DoHTML")
+	doReport(params, "syz-cover.html", rg.DoHTML)
+
+	log.Logf(1, "start DoSubsystemCover")
+	doReport(params, "syz-cover-subsystem.html", rg.DoSubsystemCover)
+
+	log.Logf(1, "start DoModuleCover")
+	doReport(params, "syz-cover-module.html", rg.DoModuleCover)
+
+	log.Logf(1, "start DoFuncCover")
+	doReport(params, "syz-cover-funccover.csv", rg.DoFuncCover)
+
+	log.Logf(1, "start DoRawCover")
+	doReport(params, "rawcoverpcs", rg.DoRawCover)
+
+	log.Logf(1, "start DoRawCoverFiles")
+	doReport(params, "rawcoverfiles", rg.DoRawCoverFiles)
+}
+
+func doReport(params cover.HandlerParams, fname string,
+	fn func(w io.Writer, params cover.HandlerParams) error) {
+	buf := new(bytes.Buffer)
+	if err := fn(buf, params); err != nil {
 		tool.Fail(err)
 	}
-	fn += ".html"
-	if err := osutil.WriteFile(fn, buf.Bytes()); err != nil {
+	log.Logf(0, "write to %v", fname)
+	if err := osutil.WriteFile(fname, buf.Bytes()); err != nil {
 		tool.Fail(err)
 	}
-	if err := exec.Command("xdg-open", fn).Start(); err != nil {
-		tool.Failf("failed to start browser: %v", err)
-	}
+	exec.Command("xdg-open", fname).Start()
 }
 
 func initPCs(rg *cover.ReportGenerator) []uint64 {
 	var pcs []uint64
 	if len(flag.Args()) == 0 {
-		for _, s := range rg.Symbols {
-			pcs = append(pcs, s.PCs...)
-		}
+		pcs = rg.CallbackPoints
```

Comment:
The motivation is not immediately clear. What are the PCs that are present in `CallbackPoints`, but we not collected in the previous approach?

Reply:
modified comment in git log.